File: december 2017/silver/measurement/measurement.1.py
"""
ID: warwick2
LANG: PYTHON3
TASK: measurement
"""
with open('measurement.in', 'r') as fin:
    in_text = fin.read().strip()
    lines = in_text.split('\n')

n, g = lines[0].split(' ')
n = int(n)
g = int(g)

entries = [] # (day, cow, change)
for line in lines[1:]:
    day, cow, change = line.split(' ')
    day = int(day)
    change = int(change)
    entries.append((day, cow, change))

entries = sorted(entries, key=lambda x: x[0])
first_day = entries[0][0]
last_day = entries[-1][0]

print(entries)

# Start all cows at 7
running_count = {}
for entry in entries:
    cow = entry[1]
    if cow in running_count:
        continue
    running_count[cow] = g

current_top = None
num_changes = 0

def pick_top():
    # super simple, look for all cows with top value
    all_cows = []
    for cow in running_count:
        all_cows.append((cow, running_count[cow]))
    all_cows = sorted(all_cows, key=lambda x: x[1], reverse=True)

    top = []
    for cow in all_cows:
        if cow[1] == all_cows[0][1]: # if cow has top score
            top.append(cow[0])
        else:
            break
    top = sorted(top)
    return top

current_top = pick_top()

print(current_top)

for day in range(first_day, last_day+1):
    for entry in entries:
        if entry[0] == day:
            # update cow on day
            cow = entry[1]
            change = entry[2]
            running_count[cow] += change
    new_top = pick_top()
    if current_top != new_top:
        num_changes += 1
        current_top = new_top

out = str(num_changes)
out += '\n'
with open('measurement.out', 'w') as fout:
    fout.write(out)

File: december 2017/silver/measurement/measurement1.cpp
/*
ID: warwick2
LANG: C++11
TASK: base
*/
#include <iostream>
#include <fstream>
#include <algorithm>
#include <set>
#include <list>
#include <cmath>
#include <vector>
using namespace std;

#define MAX_N 100000

struct entry{
    int day;
    int cow;
    int change;
};

bool sort_entries(entry a, entry b){
    return b.day > a.day;
}

int n, g;
entry entries[MAX_N];
int gallons[10000];
vector<int> best_cows;

vector<int> find_best_cows(){
    vector<int> best;
    int max = 0;
    for (int i=0; i<n; i++){
        if (gallons[i] > max){
            best.clear();
            best.emplace_back(i);
        }
        if (gallons[i] == max){
            best.emplace_back(i);
        }
    }

    for (auto a : best){
        cout << "best" << a << "\n";
        cout << "\n";
    }

    return best;
}

int main(void){
	ifstream fin("measurement.in");
    ofstream fout("measurement.out");

    fin >> n >> g;
    for (int i=0; i<n; i++){
        int day, cow, change;
        fin >> day >> cow >> change;
        entries[i] = {day, cow, change};
    }

    sort(entries, entries+n, sort_entries);

    int next_entry = 0;
    int changes = 0;
    int first_day = entries[0].day;
    int last_day = entries[n-1].day;    

    best_cows = find_best_cows();

    for (int day=first_day; day<=last_day; day++){
        while (entries[next_entry].day == day){
            gallons[entries[next_entry].cow] += entries[next_entry].change;
            next_entry++;
        }
        vector<int> new_best_cows = find_best_cows();
        if (!equal(best_cows.begin(), best_cows.begin()+n, new_best_cows.begin())){
            changes++;
            best_cows = new_best_cows;
        }
    }

    fout << changes << "\n";
    return 0;
}